### PR TITLE
Mario+Fabo+Ana/3996 polkadot unstake

### LINKED ISSUE #3996: Implement Unstake in Polkadot
**Is your feature request related to a problem? Please describe.**

Implement Unstake in Polkadot, see related extrinsic:

https://polkadot.js.org/api/substrate/extrinsics.html#unbond-value-compact-balanceof-t

### PR DESCRIPTION
**Pull request by faboweb**
Closes #3996

**Description:**

### WARNING! PLOT TWIST :-)

**Stash can:** 
- bond
- bondExtra

**Controller can:**
- nominate
- unbond (partially/all)

So actions should be:

- [x] Disable Unstake for stash accounts, we  should warn about it in some place.
- [x] No limits for unstake using a controller account

Thank you! 🚀

---

For contributor:

- [ ] Added changes entries. Run `yarn changelog` for a guided process.
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

**Comment by Bitcoinera** [comment]
It seems it works! Here the transaction: https://polkascan.io/pre/kusama/transaction/0x8b03863d09f846d7fd503f0e6803e083146b3604e26681ce63ff6f68e197f7db

**Comment by faboweb** [comment]
Sweet. Need to consider https://github.com/luniehq/lunie/pull/4058 and fix some tests I guess

**Comment by Bitcoinera** [comment]
> Sweet. Need to consider #4058 and fix some tests I guess

Ah true...

**Comment by Bitcoinera** [comment]
I guess the idea here is to add to `UndelegationModal` the same than `DelegationModal` right?:
`:is-disabled="session.addressRole === `stash`"`

**Comment by faboweb** [comment]
yep. disable validator selection if stash account the same way. and disable amount input if controller.

File: src/ActionModal/components/UndelegationModal.vue [hunk 1]
```diff
@@ -18,6 +18,30 @@
     @close="clear"
     @txIncluded="onSuccess"
   >
+    <TmFormGroup
+      v-if="session.addressRole === `stash`"
+      class="action-modal-form-group"
+    >
+      <div class="form-message notice">
+        <span>
+          This is a stash account, you can increase the amount to stake but you
```
Comment (faboweb) [inline src/ActionModal/components/UndelegationModal.vue:27]:
```suggestion
          This is a stash account, you can decrease the amount to stake but you
```

File: src/ActionModal/components/UndelegationModal.vue [hunk 2]
```diff
@@ -18,6 +18,30 @@
...
+          need to sign in with your controller account to set or change your
```
Comment (faboweb) [inline src/ActionModal/components/UndelegationModal.vue:28]:
```suggestion
          need to sign in with your controller account to change your
```

File: src/ActionModal/components/UndelegationModal.vue [hunk 3]
```diff
@@ -18,6 +18,30 @@
...
+        </span>
+      </div>
+    </TmFormGroup>
+    <TmFormGroup
+      v-if="session.addressRole === `controller`"
+      class="action-modal-form-group"
+    >
+      <div class="form-message notice">
+        <span>
+          This is a controller account, you can set or change your validators
```
Comment (faboweb) [inline src/ActionModal/components/UndelegationModal.vue:39]:
```suggestion
          This is a controller account, you can change your validators
```

File: src/ActionModal/components/UndelegationModal.vue [hunk 4]
```diff
@@ -18,6 +18,30 @@
...
+          but to increase the amount to stake you need to sign in with your
```
Comment (faboweb) [inline src/ActionModal/components/UndelegationModal.vue:40]:
```suggestion
          but to decrease the amount you stake you need to sign in with your
```

File: src/signing/networkMessages/kusama.js [hunk 5]
```diff
@@ -68,7 +72,8 @@ export async function UnstakeTx(senderAddress, { from, amount }, network) {
     transactions.push(await api.tx.staking.unbond(chainAmount))
   }
 
-  if (from.length > 0) {
+  // Disable adding possible nomination extrinsic if address is a stash account
+  if (from.length > 0 && addressRole !== `stash`) {
```
Comment (faboweb) [inline src/signing/networkMessages/kusama.js:76]:
```suggestion
  if (from.length > 0 && ["controller", "stash/controller", "none"].includes(addressRole)) {
```

inverted for clarity

File: src/signing/networkMessages/kusama.js [hunk 6]
```diff
@@ -91,8 +96,8 @@ export async function ClaimRewardsTx(senderAddress) {
   if (newStakerRewards.length === 0) {
     allClaimingTxs = []
   } else {
-    newStakerRewards.forEach((reward) => {
-      reward.nominating.forEach((nomination) => {
+    newStakerRewards.forEach(reward => {
```
Comment (faboweb) [inline src/signing/networkMessages/kusama.js:99]: update your packages and lint again

File: src/signing/networkMessages/kusama.js (hunk 5 above)
Comment (Bitcoinera) [reply src/signing/networkMessages/kusama.js:76]: should include `undefined` as well no?

File: src/ActionModal/components/UndelegationModal.vue (hunk 4 above)
Comment (Bitcoinera) [reply src/ActionModal/components/UndelegationModal.vue:40]: Copy paste 😅

File: src/signing/networkMessages/kusama.js (hunk 5 above)
Comment (Bitcoinera) [reply src/signing/networkMessages/kusama.js:76]: ah no, because it is only for Kusama. Cool

**Comment by mariopino** [comment]
@Bitcoinera succesfully tested this, so only unit tests are missing!

File: src/ActionModal/components/UndelegationModal.vue [hunk 7]
```diff
@@ -294,10 +301,15 @@ export default {
...
     return {
       amount: {
-        required: (x) => !!x && x !== `0`,
+        required: (x) =>
+          this.currentNetwork.network_type === "polkadot" || (!!x && x !== `0`),
         decimal,
-        max: (x) => Number(x) <= this.maximum,
-        min: (x) => Number(x) >= SMALLEST,
+        max: (x) =>
+          this.currentNetwork.network_type === "polkadot" ||
```
Comment (faboweb) [inline src/ActionModal/components/UndelegationModal.vue:308]: max does apply for any network, doesn't it?

File: src/components/staking/PageValidator.vue [hunk 8]
```diff
@@ -82,7 +79,7 @@
         <TmBtn
           id="undelegation-btn"
           class="undelegation-btn"
-          :disabled="delegation.amount === 0"
+          :disabled="delegation.noDelegation || session.addressRole === `stash`"
```
Comment (faboweb) [inline src/components/staking/PageValidator.vue:82]: nope! please allow and show in action modal

File: src/signing/networkMessages/kusama.js [hunk 9]
```diff
@@ -49,6 +49,7 @@ export async function StakeTx(
         delegatedValidators.concat(to[0]),
         (x) => x
       )
+      console.log(transactions)
```
Comment (faboweb) [inline src/signing/networkMessages/kusama.js:52]:
```suggestion
```

**Comment by Bitcoinera** [comment]
I feel forgotten in the title of this PR 😿

**Comment by Bitcoinera** [comment]
MadeInThePile-s30-s30studio-set30-LPxavt0E2Fd0Y4cXts

File: src/ActionModal/components/UndelegationModal.vue (hunk 7 above)
Comment (Bitcoinera) [reply src/ActionModal/components/UndelegationModal.vue:308]: Yes, I also don't understand this distinction actually

**Comment by faboweb** [comment]
Adjusted @Bitcoinera :D

File: src/components/staking/PageValidator.vue (hunk 8 above)
Comment (Bitcoinera) [reply src/components/staking/PageValidator.vue:82]: oki

**Comment by faboweb** [comment]
unit tests failing and merge needed

**Comment by faboweb** [comment]
The unstake button is still disabled for a stash account (Anas Stash and Federico)

**Comment by faboweb** [comment]
Can't test the controller as we need to merge the better validator lookup in the API first

**Comment by Bitcoinera** [comment]
> Can't test the controller as we need to merge the better validator lookup in the API first

You could use local API

**Comment by Bitcoinera** [comment]
> The unstake button is still disabled for a stash account (Anas Stash and Federico)

It works on Delega .. 👀  because according to UI I don't have any delegation on Federico, that's why

**Comment by faboweb** [comment]
> It works on Delega .. 👀 because according to UI I don't have any delegation on Federico, that's why

that's wrong. in Polkadot you can have delegations but they are currently inactive/0. But you still might want to remove those validators from your set. @mariopino might want to take care of that.

**Comment by Bitcoinera** [comment]
> that's wrong. in Polkadot you can have delegations but they are currently inactive/0. But you still might want to remove those validators from your set. @mariopino might want to take care of that.

Good point. But I think this PR got big enough. We should open an issue and do this on another PR, don't you think?

Right now, unstaking does work, but only on the currently active delegation. Adding the unstake as removal from validator set would make another PR imo

**Comment by faboweb** [comment]
I am not sure you understand Polkadot staking. Nominations are independent from bonded stake. And the actual distribution of your bonded stake across your nominations is dynamic. So your nominated validators might have stake in one session and in another they don't.

**Comment by Bitcoinera** [comment]
> I am not sure you understand Polkadot staking. Nominations are independent from bonded stake. And the actual distribution of your bonded stake across your nominations is dynamic. So your nominated validators might have stake in one session and in another they don't.

I knew this, but I still argued that the bonded amount is being in fact unstaked and that dealing with the nominations part would be another PR.

But alright, it is an easy change, so let's do it here

**Comment by faboweb** [comment]
You need a controller account. But you should always open the actionmodal and then advice the user there. Because like it is right now it is intransparent to the user why he can't unstake.

**Comment by Bitcoinera** [comment]
> You need a controller account. But you should always open the actionmodal and then advice the user there. Because like it is right now it is intransparent to the user why he can't unstake.

OK, I finally got you! I understood that you wanted to make controller able to unnominate the inactive validator. But you meant simply adding a message for stash accounts about why they can't unstake. Ok ok 👍

File: src/ActionModal/components/UndelegationModal.vue [hunk 10]
```diff
@@ -105,7 +77,6 @@
           class="tm-field-addon"
           placeholder="0"
           type="number"
-          :is-disabled="session.addressRole === `controller`"
```
Comment (Bitcoinera) [inline src/ActionModal/components/UndelegationModal.vue]: @mariopino why did you remove this? Controller accounts shouldn't be able to change the staked amount right?

File: src/ActionModal/components/UndelegationModal.vue [hunk 11]
```diff
@@ -62,11 +38,7 @@
     >
       <TmField
         id="from"
-        :value="
-          session.addressRole === `stash`
-            ? `--`
-            : sourceValidator | validatorEntry
-        "
```
Comment (Bitcoinera) [inline src/ActionModal/components/UndelegationModal.vue]: This also makes sense to me for a stash account

File: src/ActionModal/components/UndelegationModal.vue (hunk 10 above)
Comment (faboweb) [reply src/ActionModal/components/UndelegationModal.vue]: Only controllers can reduce the staked amount

Comment (Bitcoinera) [reply src/ActionModal/components/UndelegationModal.vue]: OK, I was wrong, the can and it is stash accounts who cannot do anything unstake related it seems. Just checked

**Comment by Bitcoinera** [comment]
IMO there is still something looking a bit fishy with the nominating extrinsics. According to the documentation, in `api.tx.staking.nominate(validatorAddresses)` you are selecting your target validators. So, if we exclude the current validator we are unstaking from, we are also unnominating it...

But this should only happen if we unstake ALL the delegation? Ah yeah, but Polkadot staking works in this different way.

Maybe add a checkbox? "Unnominate validator". Right now we are unnominating it by default

**Comment by Bitcoinera** [comment]
Also another question regarding this. I just made a transaction with the extrinsic to unnominate. Should be there at least. When will I see the change (validator out of my validatorsset), in next era or after the 7 days period the unlocking takes?

File: src/ActionModal/components/UndelegationModal.vue [hunk 12]
```diff
@@ -18,6 +18,18 @@
     @close="clear"
     @txIncluded="onSuccess"
   >
+    <TmFormGroup
+      v-if="session.addressRole === `stash`"
+      class="action-modal-form-group"
+    >
+      <div class="form-message notice">
+        <span>
+          This is a stash account, you can decrease the amount you stake but you
```
Comment (faboweb) [inline src/ActionModal/components/UndelegationModal.vue]: @mariopino this is wrong, right?

File: src/ActionModal/components/UndelegationModal.vue [hunk 13]
```diff
@@ -38,7 +50,12 @@
     >
       <TmField
         id="from"
-        :value="sourceValidator | validatorEntry"
+        :value="
+          // prettier-ignore
```
Comment (faboweb) [inline src/ActionModal/components/UndelegationModal.vue]: ?

File: src/signing/networkMessages/kusama.js [hunk 14]
```diff
@@ -82,7 +82,10 @@ export async function UnstakeTx(
     const validatorAddresses = delegatedValidators.filter(
       (validator) => !from.includes(validator)
     )
-    transactions.push(await api.tx.staking.nominate(validatorAddresses))
+    // transaction will fail if targets are empty
```
Comment (faboweb) [inline src/signing/networkMessages/kusama.js:87]: that is bad: that would mean you can not reduce nominations to none

File: src/ActionModal/components/UndelegationModal.vue (hunk 12 above)
Comment (Bitcoinera) [reply src/ActionModal/components/UndelegationModal.vue]: It is outdated. Now it reads "This is a stash account, you cannot perform any unstake related
          action. To decrease the amount your stake and change your validators
          you need to sign in with your controller account"

Comment (Bitcoinera) [reply src/ActionModal/components/UndelegationModal.vue]: Maybe you can change nominations though 🤔 Now thinking about it

Comment (Bitcoinera) [reply src/ActionModal/components/UndelegationModal.vue]: I am going to test

File: src/ActionModal/components/UndelegationModal.vue (hunk 13 above)
Comment (Bitcoinera) [reply src/ActionModal/components/UndelegationModal.vue]: it is also outdated. No it is a computed, this one:

```
    enhancedSourceValidator() {
      return validatorEntry(this.sourceValidator)
    },
```

The reason is because, for some reason, `validatorEntry` was throwing an awful error, don't know why. This typical error "component has not being loaded but is mentioned during rendered. Check your reactive component".

Could only fix it by doing this 🤷

File: src/signing/networkMessages/kusama.js (hunk 14 above)
Comment (Bitcoinera) [reply src/signing/networkMessages/kusama.js:87]: to do that you need to to `forceUnstake`

Comment (Bitcoinera) [reply src/signing/networkMessages/kusama.js:87]: https://polkadot.js.org/api/substrate/extrinsics.html#forceunstake-stash-t-accountid-num-slashing-spans-u32

Maybe there is some other command. @mariopino do you know?

I can confirm nominate fails when validators addresses is `[]`

Comment (Bitcoinera) [reply src/signing/networkMessages/kusama.js:87]: [Here ](https://polkascan.io/pre/kusama/transaction/0xc4204d60b987f0aa93e3ec7a11cfcbc1d711338865dc27f77735a73be9dfce0c)the failed transaction

Comment (mariopino) [reply src/signing/networkMessages/kusama.js:87]: This `api.tx.staking.chill` maybe can do the job:
https://polkadot.js.org/api/substrate/extrinsics.html#chill

It does the same thing as `forceUnstake` but effects will be felt at the beginning of the next era.

Comment (Bitcoinera) [reply src/signing/networkMessages/kusama.js]: chill man

**Comment by Bitcoinera** [comment]
> Also another question regarding this. I just made a transaction with the extrinsic to unnominate. Should be there at least. When will I see the change (validator out of my validatorsset), in next era or after the 7 days period the unlocking takes?

Look, it worked! I successfully unnominated Delega. And it happened on era change:

File: src/signing/networkMessages/kusama.js (hunk 14 above)
Comment (faboweb) [reply src/signing/networkMessages/kusama.js]: let's add the chill logic?
